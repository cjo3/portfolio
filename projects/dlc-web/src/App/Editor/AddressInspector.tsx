import { Drawer, List, ListItem, ListItemText } from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import React from 'react'
import {
  TransformResultCellCollection,
  TransformSummary
} from '../../store/editor/interfaces'

const useStyles = makeStyles(theme => ({
  TransformResults_sideDrawer: {
    ...theme.custom.setFlex('column'),
    padding: theme.custom.setSpace(),
    backgroundColor: theme.palette.grey[900]
  },
  TransformResults_innerList: {
    height: '100%'
  },
  TransformResults_addressListItem: {
    ...theme.custom.noSelect,
    'transition': 'all 250ms linear',
    'color': theme.palette.grey[400],
    '&:hover': {
      color: theme.palette.grey[50]
    }
  }
}))

interface Props {
  drawerOpen: boolean
  closeDrawerHandler: (event: React.MouseEvent) => void
  drawerDataName: string
  allTransforms: TransformResultCellCollection
  addresses: TransformSummary[]
}

export const AddressInspector: React.FC<Props> = ({
  drawerOpen,
  closeDrawerHandler,
  drawerDataName,
  allTransforms,
  addresses
}): JSX.Element => {
  const classes = useStyles()
  if (!addresses) return null
  return (
    <Drawer
      anchor="left"
      open={drawerOpen}
      onClose={closeDrawerHandler}
      PaperProps={{
        classes: {
          root: classes.TransformResults_sideDrawer
        }
      }}>
      <List className={classes.TransformResults_innerList}>
        {drawerDataName &&
          addresses.map(address => (
            <ListItem
              key={`drawer-data-item-${address}`}
              className={classes.TransformResults_addressListItem}
              alignItems="center">
              <ListItemText
                primary={address}
                primaryTypographyProps={{
                  align: 'center',
                  variant: 'h5'
                }}
                secondary={`${allTransforms[address].original} >> ${allTransforms[address].result.w}`}
                secondaryTypographyProps={{
                  align: 'center',
                  variant: 'body1',
                  color: 'secondary'
                }}
              />
            </ListItem>
          ))}
      </List>
    </Drawer>
  )
}
