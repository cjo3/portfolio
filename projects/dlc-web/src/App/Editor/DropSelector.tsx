import { makeStyles } from '@material-ui/core/styles'
import GetAppIcon from '@material-ui/icons/GetApp'
import clsx from 'clsx'
import path from 'path'
import React, { useState } from 'react'
import { FileUpload } from '../../interfaces'

const useStyles = makeStyles(theme => ({
  dropContainer: {
    ...theme.custom.setFlex('column'),
    width: '100%',
    height: 200,
    borderRadius: theme.custom.setSpace() / 4,
    backgroundColor: theme.palette.primary[300],
    transition: 'all 250ms linear'
  },
  dragColor: {
    boxShadow: theme.custom.boxShadow,
    backgroundColor: theme.palette.primary[400]
  },
  dropIcon: {
    color: theme.palette.grey[50],
    fontSize: theme.typography.fontSize * 5
  }
}))

interface Props {
  setSelectedFileHandler: (file: FileUpload) => void
}

export const DropSelector: React.FC<Props> = ({
  setSelectedFileHandler
}): JSX.Element => {
  const classes = useStyles()

  const [dragCount, setDragCount] = useState<number>(0)

  const dragEnterHandler = (event: React.DragEvent<HTMLDivElement>): void => {
    event.preventDefault()
    setDragCount(dragCount + 1)
  }

  const dragLeaveHandler = (event: React.DragEvent<HTMLDivElement>): void => {
    event.preventDefault()
    setDragCount(dragCount - 1)
  }

  const dragOverHandler = (event: React.DragEvent<HTMLDivElement>): void => {
    event.preventDefault()
  }

  const dropHandler = (event: React.DragEvent<HTMLDivElement>): void => {
    event.preventDefault()
    const item = event.dataTransfer.items[0]
    if (item.kind === 'file') {
      const file = item.getAsFile()
      const fileExt = path.extname(file.name).toLowerCase()
      if (process.env.ACCEPTED_FILETYPES.split(',').includes(fileExt))
        setSelectedFileHandler(file)
    }
    setDragCount(0)
  }

  return (
    <div
      data-testid="DropSelector"
      className={clsx(classes.dropContainer, {
        [classes.dragColor]: dragCount > 0
      })}
      onDragEnter={dragEnterHandler}
      onDragLeave={dragLeaveHandler}
      onDragOver={dragOverHandler}
      onDrop={dropHandler}>
      <GetAppIcon className={classes.dropIcon} />
    </div>
  )
}
