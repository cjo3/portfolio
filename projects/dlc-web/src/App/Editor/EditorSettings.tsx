import clsx from 'clsx'
import {
  Grid,
  Typography,
  TextField,
  FormControlLabel,
  Radio,
  RadioGroup,
  FormControl,
  FormLabel,
  ButtonGroup,
  Button
} from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import React from 'react'

const useStyles = makeStyles(theme => ({
  subSectionContainer: {
    marginTop: theme.custom.setSpace('sm')
  },
  textWhite: {
    color: theme.palette.grey[50]
  },
  settingsPanel: {
    borderSize: 1,
    padding: theme.custom.setSpace()
  },
  scopeRangePanel: {
    background: `linear-gradient(top, ${theme.palette.grey[100]}, white)`
  },
  underLimitPanel: {
    background: `linear-gradient(top, ${theme.palette.primary[100]}, white)`
  },
  overLimitPanel: {
    background: `linear-gradient(top, ${theme.palette.secondary[100]}, white)`
  },
  topMargin: {
    marginTop: theme.custom.setSpace()
  },
  fieldSetMargin: {
    marginTop: theme.custom.setSpace() * 1.5
  },
  radioGroupLabel: {
    ...theme.typography.bold,
    marginBottom: theme.custom.setSpace() / 2
  },
  fontSize: {
    fontSize: theme.typography.fontSize
  },
  noMargin: {
    margin: 0
  }
}))

export const EditorSettings: React.FC = ({}): JSX.Element => {
  const classes = useStyles()

  return (
    <Grid item xs={12}>
      <Grid container>
        <Grid item xs={12} className={classes.textWhite}>
          <Typography variant="h3">Editor Settings</Typography>
          <Typography variant="body1">
            Tempor dolor amet at dolor no tempor ipsum tempor et diam. Justo
            sanctus lorem voluptua vero sit. Erat vero sed et duo elitr diam,
            clita et rebum amet magna. Amet ut takimata vero aliquyam dolores ea
            ea, justo eirmod vero kasd labore justo consetetur, ea accusam elitr
            elitr sit gubergren.
          </Typography>
        </Grid>
      </Grid>
      <Grid container className={classes.subSectionContainer}>
        <Grid
          item
          xs={12}
          sm={4}
          className={clsx(classes.settingsPanel, classes.scopeRangePanel)}>
          <Typography variant="h5">Editor Scope</Typography>
          <TextField
            fullWidth
            label="Range Start"
            id="range-start"
            variant="outlined"
            size="small"
            className={classes.topMargin}
            placeholder="C5"
            InputProps={{
              classes: {
                input: classes.fontSize
              }
            }}
          />
          <TextField
            fullWidth
            label="Range End"
            id="range-end"
            variant="outlined"
            size="small"
            className={classes.topMargin}
            placeholder="S63"
            InputProps={{
              classes: {
                input: classes.fontSize
              }
            }}
          />
          <Typography variant="body1">
            All cells will be checked if unset
          </Typography>
        </Grid>
        <Grid
          item
          xs={12}
          sm={4}
          className={clsx(classes.settingsPanel, classes.underLimitPanel)}>
          <Typography variant="h5">Under Limit</Typography>
          <TextField
            fullWidth
            required
            label="Trigger Character"
            id="ul-trigger-char"
            variant="outlined"
            size="small"
            className={classes.topMargin}
            placeholder="<, -, other"
            defaultValue="<"
            InputProps={{
              classes: {
                input: classes.fontSize
              }
            }}
          />
          <TextField
            fullWidth
            label="Zero Trigger"
            id="ul-zero-trigger"
            variant="outlined"
            size="small"
            className={classes.topMargin}
            placeholder="null, n/a, void, etc."
            InputProps={{
              classes: {
                input: classes.fontSize
              }
            }}
          />
          <FormControl
            component="fieldset"
            className={classes.fieldSetMargin}
            required>
            <FormLabel
              component="legend"
              className={clsx(classes.radioGroupLabel, classes.fontSize)}>
              Function Select
            </FormLabel>
            <RadioGroup aria-label="ul-function-select" name="ul-function">
              <FormControlLabel
                classes={{
                  root: classes.noMargin,
                  label: classes.noMargin
                }}
                value="leave"
                control={<Radio size="small" color="primary" />}
                label="Leave Value"
              />
              <FormControlLabel
                classes={{
                  root: classes.noMargin,
                  label: classes.noMargin
                }}
                value="halve"
                control={<Radio size="small" color="primary" />}
                label="Halve Value"
              />
              <FormControlLabel
                classes={{
                  root: classes.noMargin,
                  label: classes.noMargin
                }}
                value="zero"
                control={<Radio size="small" color="primary" />}
                label="Zero Value"
              />
              <FormControlLabel
                classes={{
                  root: classes.noMargin,
                  label: classes.noMargin
                }}
                disabled
                value="custom-value"
                control={<Radio size="small" color="primary" />}
                label="Custom Function"
              />
            </RadioGroup>
          </FormControl>
        </Grid>
        <Grid
          item
          xs={12}
          sm={4}
          className={clsx(classes.settingsPanel, classes.overLimitPanel)}>
          <Typography variant="h5">Over Limit</Typography>
          <TextField
            fullWidth
            required
            label="Trigger Character"
            id="ol-trigger-char"
            variant="outlined"
            size="small"
            color="secondary"
            className={classes.topMargin}
            placeholder=">, +, other"
            defaultValue=">"
            InputProps={{
              classes: {
                input: classes.fontSize
              }
            }}
          />
          <TextField
            fullWidth
            label="Over Trigger"
            id="ol-over-trigger"
            variant="outlined"
            size="small"
            className={classes.topMargin}
            color="secondary"
            placeholder="over, high, exceeded etc."
            InputProps={{
              classes: {
                input: classes.fontSize
              }
            }}
          />
          <FormControl
            component="fieldset"
            className={classes.fieldSetMargin}
            required>
            <FormLabel
              component="legend"
              color="secondary"
              className={clsx(classes.radioGroupLabel, classes.fontSize)}>
              Function Select
            </FormLabel>
            <RadioGroup aria-label="ol-function-select" name="ol-function">
              <FormControlLabel
                classes={{
                  root: classes.noMargin,
                  label: classes.noMargin
                }}
                value="leave"
                control={<Radio size="small" />}
                label="Leave Value"
              />
              <FormControlLabel
                classes={{
                  root: classes.noMargin,
                  label: classes.noMargin
                }}
                value="halve"
                control={<Radio size="small" />}
                label="Halve Value"
              />
              <FormControlLabel
                classes={{
                  root: classes.noMargin,
                  label: classes.noMargin
                }}
                value="zero"
                control={<Radio size="small" />}
                label="Zero Value"
              />
              <FormControlLabel
                classes={{
                  root: classes.noMargin,
                  label: classes.noMargin
                }}
                disabled
                value="custom-value"
                control={<Radio size="small" />}
                label="Custom Function"
              />
            </RadioGroup>
          </FormControl>
        </Grid>
      </Grid>
      <Grid container className={classes.subSectionContainer}>
        <Grid item xs={6}>
          <Grid container direction="column">
            <Typography
              variant="h5"
              align="center"
              className={classes.textWhite}>
              Sheet Selection
            </Typography>
            <ButtonGroup
              className={classes.topMargin}
              orientation="vertical"
              color="primary"
              aria-label="vertical outlined primary button group">
              <Button>One</Button>
              <Button>Two</Button>
              <Button>Three</Button>
            </ButtonGroup>
          </Grid>
        </Grid>
        <Grid item xs={6}>
          <Grid
            container
            direction="column"
            justify="center"
            alignItems="center">
            <Typography
              variant="h5"
              align="center"
              className={classes.textWhite}>
              Next Actions
            </Typography>
            <ButtonGroup
              variant="outlined"
              color="secondary"
              className={classes.topMargin}
              aria-label="contained primary button group">
              <Button>Submit</Button>
              <Button>Reset</Button>
            </ButtonGroup>
          </Grid>
        </Grid>
      </Grid>
    </Grid>
  )
}
