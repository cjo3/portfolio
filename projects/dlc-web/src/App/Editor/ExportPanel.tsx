import { writeWorkbook } from '@cjo3/shared/logic/dlc'
import {
  currentSheetName as currentSheetNameMock,
  sheetData as sheetDataMock,
  transformResult as transformResultMock,
  workbookName as workbookNameMock
} from '@cjo3/shared/react/mocks/dlc'
import { Button, ButtonGroup, Grid, Typography } from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import { Elements } from '@stripe/react-stripe-js'
import { loadStripe } from '@stripe/stripe-js'
import React, { useState } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { exportButtons } from '../../constants'
import { openSnackbar } from '../../store/app/actions'
import { ExportData } from '../../store/editor/interfaces'
import {
  currentSheetNameSelector,
  sheetDataSelector,
  transformResultSelector,
  workbookNameSelector
} from '../../store/selectors'
import { PaymentDialog } from './PaymentDialog/'

const stripePromise = loadStripe(process.env.STRIPE_PUBLIC_KEY)

const useStyles = makeStyles(theme => ({
  ExportPanel_container: {
    ...theme.custom.borderRadius,
    'padding': theme.custom.setSpace('sm'),
    'background': theme.custom.setLinearGradient(
      180,
      theme.palette.secondary.main,
      theme.palette.secondary[300]
    ),
    'color': theme.palette.secondary[50],
    'transition': 'all 250ms ease-out',
    [theme.breakpoints.down('xs')]: {
      marginTop: theme.custom.setSpace('sm')
    },
    '&:hover': {
      boxShadow: theme.custom.boxShadow
    }
  },
  ExportPanel_topMargin: {
    marginTop: theme.custom.setSpace('sm')
  }
}))

export const ExportPanel: React.FC = (): JSX.Element => {
  const classes = useStyles()

  let workbookName = useSelector(workbookNameSelector)
  let currentSheetName = useSelector(currentSheetNameSelector)
  let sheetData = useSelector(sheetDataSelector)
  let transformResult = useSelector(transformResultSelector)

  if (process.env.USE_MOCKS) {
    workbookName = workbookNameMock
    currentSheetName = currentSheetNameMock
    sheetData = sheetDataMock
    transformResult = transformResultMock
  }

  const dispatch = useDispatch()

  const [paymentOpen, setPaymentOpen] = useState<boolean>(false)
  const [exportData, setExportData] = useState<ExportData | null>(null)

  const openPaymentHandler = (
    event: React.MouseEvent<HTMLButtonElement>
  ): void => {
    try {
      const wbBuffer = writeWorkbook(
        sheetData,
        transformResult.all,
        currentSheetName,
        event.currentTarget.dataset.booktype
      )

      const fileName = `${workbookName}-edited.${event.currentTarget.name}`

      const blob: Blob = new Blob([wbBuffer], {
        type: 'application/octet-stream'
      })

      setExportData({
        fileName,
        blob
      })
      setPaymentOpen(true)
    } catch (error) {
      dispatch(
        openSnackbar(
          'Sorry, file export cannot be processed at this time',
          'error'
        )
      )
    }
  }

  const closePaymentHandler = (): void => {
    setExportData(null)
    setPaymentOpen(false)
  }

  return (
    <Grid item xs={12} sm={6} className={classes.ExportPanel_container}>
      {paymentOpen && (
        <Elements stripe={stripePromise}>
          <PaymentDialog
            open={paymentOpen}
            closeHandler={closePaymentHandler}
            exportData={exportData}
          />
        </Elements>
      )}
      <Typography variant="h5">
        Export for only ${process.env.EXPORT_PRICE}
      </Typography>
      <Typography variant="body1">
        Download your transformed sheet after purchase by selecting a file type
        below. If your sheet cannot be exported, we will let you know before
        checkout. Also, we do not save copies of exported sheets so please be
        careful with them after download!
      </Typography>
      <ButtonGroup className={classes.ExportPanel_topMargin}>
        {exportButtons.map(item => (
          <Button
            key={item.key}
            type="button"
            name={item.name}
            data-booktype={item.bookType}
            onClick={openPaymentHandler}>
            {item.label}
          </Button>
        ))}
      </ButtonGroup>
    </Grid>
  )
}
