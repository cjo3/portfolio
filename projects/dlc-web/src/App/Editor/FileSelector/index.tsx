import { Button, Grid, Typography } from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import React, { useState } from 'react'

const useStyles = makeStyles(theme => ({
  form: {
    marginTop: theme.custom.setSpace()
  },
  hidden: {
    visibility: 'hidden'
  },
  fileName: {
    margin: `0 0 0 ${theme.custom.setSpace()}px`
  },
  formButton: {
    'marginRight': theme.custom.setSpace(),
    '&:last-of-type': {
      marginRight: 0
    }
  }
}))

interface FileObject {
  name: string
}

export const FileSelector: React.FC = (): JSX.Element => {
  const classes = useStyles()

  const initialFile: FileObject = {
    name: ''
  }

  const [selectedFile, setSelectedFile] = useState<FileObject | null>(null)

  const [savedFile, setSavedFile] = useState<FileObject | null>(null)

  const submitHandler = (event: React.FormEvent<HTMLFormElement>): void => {
    event.preventDefault()
    setSavedFile(selectedFile)
    event.currentTarget.reset()
  }

  const changeHandler = (event: React.ChangeEvent<HTMLInputElement>): void => {
    setSelectedFile(event.currentTarget.files[0])
  }

  const resetHandler = (event: React.FormEvent<HTMLFormElement>): void => {
    event.currentTarget.reset()
    setSelectedFile(initialFile)
  }

  return (
    <form
      onSubmit={submitHandler}
      onReset={resetHandler}
      className={classes.form}>
      <Grid container justify="flex-start" alignItems="center">
        <Button type="button" variant="outlined">
          <label htmlFor="file-upload-input">Select File</label>
        </Button>
        <Typography variant="body1" className={classes.fileName}>
          {selectedFile ? selectedFile.name : 'No File Selected'}
        </Typography>
      </Grid>
      <input
        className={classes.hidden}
        id="file-upload-input"
        name="file-upload-input"
        type="file"
        accept=".csv, .xls, .xlsx"
        onChange={changeHandler}
      />
      <Grid container justify="flex-start" alignItems="center">
        <Button
          type="submit"
          variant="contained"
          color="primary"
          className={classes.formButton}>
          Load
        </Button>
        <Button type="reset" variant="contained" className={classes.formButton}>
          Reset
        </Button>
      </Grid>
    </form>
  )
}

// UploadForm.propTypes = {
//   name: PropTypes.string.isRequired,
//   size: PropTypes.number.isRequired
// }
