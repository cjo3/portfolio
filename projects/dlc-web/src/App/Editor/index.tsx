import { Grid, Button, Typography } from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import React from 'react'
import { FileLoader } from './FileLoader'
import { SheetSelector } from './SheetSelector'
import { TransformResults } from './TransformResults'
import { TransformSettings } from './TransformSettings'
import { Link } from 'react-router-dom'
import { switchLinkRoutePath } from '@cjo3/shared/react/helpers'

const useStyles = makeStyles(theme => ({
  Editor_header: {
    width: '100%',
    padding: theme.custom.setSpace('lg'),
    backgroundImage:
      'url(https://eskipaper.com/images/high-tech-wallpapers-3.jpg)',
    backgroundPosition: 'center',
    backgroundSize: 'cover',
    backgroundRepeat: 'no-repeat',
    color: theme.palette.grey[50],
    textShadow: theme.custom.textShadow,
    textAlign: 'center'
  },
  Editor_header_instructionsButton: {
    marginTop: theme.custom.setSpace('sm')
  }
}))

export const Editor: React.FC = (): JSX.Element => {
  const classes = useStyles()

  return (
    <Grid container>
      <Grid item component="header" className={classes.Editor_header}>
        <Typography variant="h1" align="center">
          Editor
        </Typography>
        <Typography variant="body1" align="center">
          Labore labore ut est sit sanctus stet et. Eos lorem ipsum consetetur
          magna est voluptua. Lorem justo dolor eirmod est est aliquyam. Dolor
          sit kasd nonumy lorem tempor dolor no duo et, invidunt gubergren rebum
          at rebum, et nonumy lorem lorem diam ipsum, sadipscing diam voluptua
          accusam sit sit, dolor.
        </Typography>
        <Button
          type="button"
          variant="contained"
          color="secondary"
          className={classes.Editor_header_instructionsButton}
          href={switchLinkRoutePath(
            '/editor/instructions',
            `${process.env.APP_ROOT_PATH}/editor/instructions`
          )}>
          Instructions
        </Button>
      </Grid>
      <FileLoader />
      <SheetSelector />
      <TransformSettings />
      <TransformResults />
    </Grid>
  )
}
