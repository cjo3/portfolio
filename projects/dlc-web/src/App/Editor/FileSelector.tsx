import { Button, Hidden, Typography } from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import clsx from 'clsx'
import React, { useRef, useState } from 'react'
import { FileUpload } from '../../interfaces'
import { DropSelector } from './DropSelector'

const useStyles = makeStyles(theme => ({
  form: {
    width: '100%',
    ...theme.custom.setGrid(2, 3, theme.custom.setSpace()),
    gridTemplateRows: 'repeat(3, auto)'
  },
  title: {
    gridColumn: '1 / 2',
    gridRow: 1,
    [theme.breakpoints.down('xs')]: {
      gridColumn: '1 / 3',
      justifyContent: 'center',
      textAlign: 'center'
    }
  },
  fileSelect: {
    ...theme.custom.setFlex('row', 'flex-start'),
    gridColumn: '1 / 2',
    gridRow: 2,
    [theme.breakpoints.down('xs')]: {
      gridColumn: '1 / 3',
      justifyContent: 'center'
    }
  },
  dropLoader: {
    gridColumn: '2 / 3',
    gridRow: '1 / 4',
    ...theme.custom.setFlex(),
    [theme.breakpoints.down('xs')]: {
      display: 'none'
    }
  },
  actionButtons: {
    ...theme.custom.setFlex('row', 'flex-start'),
    gridColumn: '1 / 2',
    gridRow: 3,
    [theme.breakpoints.down('xs')]: {
      gridColumn: ' 1 / 3',
      justifyContent: 'center'
    }
  },
  hiddenInput: {
    visibility: 'hidden',
    display: 'inline-block',
    width: 1
  },
  fileName: {
    margin: `0 0 0 ${theme.custom.setSpace()}px`
  },
  formButton: {
    'marginRight': theme.custom.setSpace(),
    '&:last-of-type': {
      marginRight: 0
    }
  },
  submitButton: {
    color: 'white'
  },
  resetButton: {
    color: theme.palette.grey[600]
  }
}))

interface Props {
  setLoadedFile: (file: FileUpload) => void
}

export const FileSelector: React.FC<Props> = ({
  setLoadedFile
}): JSX.Element => {
  const classes = useStyles()

  const initialFile: FileUpload = {
    name: ''
  }

  const [selectedFile, setSelectedFile] = useState<FileUpload | null>(null)

  const changeHandler = (): void => setSelectedFile(fileInput.current.files[0])

  const submitHandler = (event: React.FormEvent<HTMLFormElement>): void => {
    event.preventDefault()
    setLoadedFile(selectedFile)
    event.currentTarget.reset()
  }

  const resetHandler = (event: React.FormEvent<HTMLFormElement>): void => {
    event.preventDefault()
    setSelectedFile(initialFile)
  }

  const fileInput = useRef<HTMLInputElement>(null)

  return (
    <form
      onSubmit={submitHandler}
      onReset={resetHandler}
      className={classes.form}>
      <div className={classes.title}>
        <Typography variant="h3">File Loader</Typography>
        <Typography variant="body1">
          Select a file using the button below
          <Hidden xsDown>
            &nbsp;
            <span>
              or drag and drop a file on the attchment box to the right
            </span>
          </Hidden>
          . The selected file name will be displayed once selected.
        </Typography>
      </div>
      <div className={classes.fileSelect}>
        <Button type="button" variant="outlined" color="primary">
          <label htmlFor="file-upload-input">Select File</label>
        </Button>
        <Typography variant="body1" className={classes.fileName}>
          {selectedFile?.size ? selectedFile.name : 'No File Selected'}
        </Typography>
        <input
          className={classes.hiddenInput}
          id="file-upload-input"
          name="file-upload-input"
          type="file"
          accept={process.env.ACCEPTED_FILETYPES}
          onChange={changeHandler}
          ref={fileInput}
        />
      </div>
      <div className={classes.dropLoader}>
        <DropSelector setSelectedFileHandler={setSelectedFile} />
      </div>
      <div className={classes.actionButtons}>
        <Button
          type="submit"
          variant="contained"
          color="primary"
          disabled={!selectedFile?.size}
          className={clsx(classes.formButton, classes.submitButton)}>
          Load
        </Button>
        <Button
          type="reset"
          variant="contained"
          disabled={!selectedFile?.size}
          className={clsx(classes.formButton, classes.resetButton)}>
          Reset
        </Button>
      </div>
    </form>
  )
}
