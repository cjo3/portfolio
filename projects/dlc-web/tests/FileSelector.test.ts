import { testRenderer, renderSnapshot } from './setup'
import { FileSelector } from '../src/App/Editor/FileSelector'

const props = {
  setLoadedFile: jest.fn()
}

describe('FileSelector', () => {
  describe('render', () => {
    test('matches current snapshop', () => {
      const tree = renderSnapshot('/editor', FileSelector, props)
      expect(tree).toMatchSnapshot()
    })

    test('renders 3 buttons', () => {
      const { getAllByRole } = testRenderer('/editor', FileSelector)
      const buttons = getAllByRole('button')
      expect(buttons.length).toBe(3)
    })

    test('renders hidden file input', () => {
      const { getByLabelText } = testRenderer('/editor', FileSelector)
      const input = getByLabelText('select file', { exact: false })
      expect(input).not.toBeVisible()
      expect(input.name).toBe('file-upload-input')
      expect(input).toHaveAttribute('type', 'file')
    })

    test('renders no file status', () => {
      const { getByText } = testRenderer('/editor', FileSelector)
      const status = getByText('no file selected', { exact: false })
      expect(status).toBeInTheDocument()
    })

    test('submit is disabled', () => {
      const { getByRole, screen } = testRenderer('/editor', FileSelector)
      const submitButton = getByRole('button', { name: /load/i })
      expect(submitButton).toHaveAttribute('disabled')
    })

    test('reset is disabled', () => {
      const { getByRole, screen } = testRenderer('/editor', FileSelector)
      const resetButton = getByRole('button', { name: /reset/i })
      expect(resetButton).toHaveAttribute('disabled')
    })
  })

  describe('events', () => {
    test('choose file', () => {
      const {
        getByLabelText,
        getByText,
        getByRole,
        userEvent: { upload }
      } = testRenderer('/editor', FileSelector)
      const input = getByLabelText('select file', { exact: false })
      const testFile = new File(['asdf'], 'asdf.xls', { type: 'xls ' })
      upload(input, testFile)
      const fileNameStatus = getByText(testFile.name, { exact: false })
      expect(fileNameStatus).toBeInTheDocument()
      const submitButton = getByRole('button', { name: /load/i })
      expect(submitButton).not.toHaveAttribute('disabled')
      const resetButton = getByRole('button', { name: /reset/i })
      expect(resetButton).not.toHaveAttribute('disabled')
    })

    test('choose file and then another', () => {
      const {
        getByLabelText,
        getByText,
        queryByText,
        userEvent: { upload }
      } = testRenderer('/editor', FileSelector)
      const input = getByLabelText('select file', { exact: false })
      const testFile = new File(['asdf'], 'asdf.xls', { type: 'xls ' })
      upload(input, testFile)
      let fileNameStatus = getByText(testFile.name, { exact: false })
      expect(fileNameStatus).toBeInTheDocument()
      const testFile2 = new File(['asdf'], 'qwer.xls', { type: 'xls ' })
      upload(input, testFile2)
      fileNameStatus = queryByText(testFile.name, { exact: false })
      expect(fileNameStatus).not.toBeInTheDocument()
      const fileNameStatus2 = getByText(testFile2.name, { exact: false })
      expect(fileNameStatus2).toBeInTheDocument()
    })

    test('choose file and reset form', () => {
      const {
        getByLabelText,
        getByRole,
        queryByText,
        userEvent: { upload, click }
      } = testRenderer('/editor', FileSelector)
      const input = getByLabelText('select file', { exact: false })
      const testFile = new File(['asdf'], 'asdf.xls', { type: 'xls ' })
      upload(input, testFile)
      const resetButton = getByRole('button', { name: /reset/i })
      click(resetButton)
      const noFileStatus = queryByText('no file selected', { exact: false })
      expect(noFileStatus).toBeInTheDocument()
      expect(resetButton).toHaveAttribute('disabled')
      const submitButton = getByRole('button', { name: /load/i })
      expect(submitButton).toHaveAttribute('disabled')
    })

    test('choose file and submit form', () => {
      const {
        getByLabelText,
        getByRole,
        userEvent: { upload, click },
        screen
      } = testRenderer('/editor', FileSelector, props)
      const input = getByLabelText('select file', { exact: false })
      const testFile = new File(['asdf'], 'asdf.xls', { type: 'xls ' })
      upload(input, testFile)
      const loadButton = getByRole('button', { name: /load/i })
      click(loadButton)
      expect(props.setLoadedFile).toHaveBeenCalled()
    })
  })
})
