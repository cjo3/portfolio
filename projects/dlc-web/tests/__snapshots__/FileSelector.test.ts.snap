// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileSelector render matches current snapshop 1`] = `
<form
  className="makeStyles-form-1"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="makeStyles-title-2"
  >
    <h3
      className="MuiTypography-root MuiTypography-h3"
    >
      File Loader
    </h3>
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      Select a file using the button below
      . The selected file name will be displayed once selected.
    </p>
  </div>
  <div
    className="makeStyles-fileSelect-3"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <label
          htmlFor="file-upload-input"
        >
          Select File
        </label>
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
    <p
      className="MuiTypography-root makeStyles-fileName-7 MuiTypography-body1"
    >
      No File Selected
    </p>
    <input
      className="makeStyles-hiddenInput-6"
      id="file-upload-input"
      name="file-upload-input"
      onChange={[Function]}
      type="file"
    />
  </div>
  <div
    className="makeStyles-dropLoader-4"
  >
    <div
      className="makeStyles-dropContainer-11"
      data-testid="drop-selector"
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root makeStyles-dropIcon-13"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 9h-4V3H9v6H5l7 7 7-7zM5 18v2h14v-2H5z"
        />
      </svg>
    </div>
  </div>
  <div
    className="makeStyles-actionButtons-5"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-formButton-8 makeStyles-submitButton-9 MuiButton-containedPrimary Mui-disabled Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="submit"
    >
      <span
        className="MuiButton-label"
      >
        Load
      </span>
    </button>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-formButton-8 makeStyles-resetButton-10 Mui-disabled Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="reset"
    >
      <span
        className="MuiButton-label"
      >
        Reset
      </span>
    </button>
  </div>
</form>
`;
